File: common/src/compat.rs
//! Client API compatibility wire contract (RFC-022 S3).
//!
//! Shared between the node's version-enforcement layer and every client
//! that must recognize a version rejection. Version values are the
//! numeric CalVer codes from [`crate::version`].

use serde::{Deserialize, Serialize};
use typeshare::typeshare;

/// Header carrying the client's identity (its CalVer code) on every
/// request to a DeviceToken surface: `x-hopnet-client-version: 20260802`.
/// Identity only — acceptance policy lives at each end (RFC-022).
pub const CLIENT_VERSION_HEADER: &str = "x-hopnet-client-version";

/// Body of a `426 Upgrade Required` version rejection. Distinct from
/// 401/403 so a version rejection never reads as a credential problem;
/// the fields tell the operator exactly what to upgrade to.
#[derive(Debug, Deserialize, Serialize)]
#[typeshare]
pub struct UpgradeRequiredResponse {
    /// The surface that rejected the request (its mount prefix).
    pub surface: String,
    /// Oldest client version code this surface accepts.
    pub min_client: u32,
    /// The node's own version code.
    pub node_version: u32,
}

File: android/HopDrive/app/src/main/java/app/hopnet/drive/net/Compat.kt
package app.hopnet.drive.net

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Client API compatibility wire contract (RFC-023 S3), mirroring
 * common/src/compat.rs. Hand-written: the app's DTOs live here rather than
 * through typeshare (see Dto.kt).
 */

/**
 * Header carrying the client's identity (its CalVer code) on every request
 * to a DeviceToken surface: `x-hopnet-client-version: 20260804`. Identity
 * only — acceptance policy lives at each end.
 */
const val CLIENT_VERSION_HEADER = "x-hopnet-client-version"

/**
 * Body of a `426 Upgrade Required` version rejection. Distinct from 401/403
 * so a version rejection never reads as a credential problem; the fields
 * tell the user exactly what to upgrade to.
 */
@Serializable
data class UpgradeRequiredResponse(
    /** The surface that rejected the request (its mount prefix). */
    val surface: String,
    /** Oldest client version code this surface accepts. */
    @SerialName("min_client") val minClient: Int,
    /** The node's own version code. */
    @SerialName("node_version") val nodeVersion: Int,
)

/** Render a CalVer code back to its display form (20260804 → "2026.8.4"). */
fun formatVersionCode(code: Int): String =
    "${code / 10_000}.${(code / 100) % 100}.${code % 100}"
